std_numeric_limits_test does not work with -O0

#### Summary:
Our tests normally are compiled with -O3, which optimizes a bunch of undefined symbols away. When compiled with -O0, just one of these tests fail, std_numeric_limits_test (both fwd and mix mode, granted we use neither of these modes).



#### Reproducible Steps:
```
cd math
echo O=0 >> make/local
./runTests.py test/unit/math/fwd/core/std_numeric_limits_test.cpp 
```

#### Current Output:
```
g++ -Wall -I . -isystem lib/eigen_3.3.3 -isystem lib/boost_1.66.0 -isystem lib/cvodes-3.1.0/include -isystem lib/idas-2.1.0/include -std=c++1y -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -Wno-unused-function -Wno-uninitialized -DGTEST_USE_OWN_TR1_TUPLE -DGTEST_HAS_PTHREAD=0 -isystem lib/gtest_1.7.0/include -isystem lib/gtest_1.7.0 -O0 -DGTEST_USE_OWN_TR1_TUPLE -DGTEST_HAS_PTHREAD=0 -isystem lib/gtest_1.7.0/include -isystem lib/gtest_1.7.0 -O0 -DNO_FPRINTF_OUTPUT -pipe  -c -o test/unit/math/fwd/core/std_numeric_limits_test.o test/unit/math/fwd/core/std_numeric_limits_test.cpp
g++ -Wall -I . -isystem lib/eigen_3.3.3 -isystem lib/boost_1.66.0 -isystem lib/cvodes-3.1.0/include -isystem lib/idas-2.1.0/include -std=c++1y -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -Wno-unused-function -Wno-uninitialized -DGTEST_USE_OWN_TR1_TUPLE -DGTEST_HAS_PTHREAD=0 -isystem lib/gtest_1.7.0/include -isystem lib/gtest_1.7.0 -O0 -DNO_FPRINTF_OUTPUT -pipe -lpthread  -o test/unit/math/fwd/core/std_numeric_limits_test test/unit/math/fwd/core/std_numeric_limits_test.o lib/gtest_1.7.0/src/gtest_main.cc lib/gtest_1.7.0/src/gtest-all.o
test/unit/math/fwd/core/std_numeric_limits_test.o: In function `AgradFwdNumericLimits_All_Fvar_Test::TestBody()':
std_numeric_limits_test.cpp:(.text+0x2e): undefined reference to `std::numeric_limits<stan::math::fvar<double> >::is_specialized'
std_numeric_limits_test.cpp:(.text+0xfd): undefined reference to `std::numeric_limits<stan::math::fvar<stan::math::fvar<double> > >::is_specialized'
std_numeric_limits_test.cpp:(.text+0x60c): undefined reference to `std::numeric_limits<stan::math::fvar<double> >::is_signed'
std_numeric_limits_test.cpp:(.text+0x6db): undefined reference to `std::numeric_limits<stan::math::fvar<stan::math::fvar<double> > >::is_signed'
std_numeric_limits_test.cpp:(.text+0x149e): undefined reference to `std::numeric_limits<stan::math::fvar<double> >::has_infinity'
std_numeric_limits_test.cpp:(.text+0x156d): undefined reference to `std::numeric_limits<stan::math::fvar<stan::math::fvar<double> > >::has_infinity'
std_numeric_limits_test.cpp:(.text+0x163c): undefined reference to `std::numeric_limits<stan::math::fvar<double> >::has_quiet_NaN'
std_numeric_limits_test.cpp:(.text+0x170b): undefined reference to `std::numeric_limits<stan::math::fvar<stan::math::fvar<double> > >::has_quiet_NaN'
std_numeric_limits_test.cpp:(.text+0x17da): undefined reference to `std::numeric_limits<stan::math::fvar<double> >::has_signaling_NaN'
std_numeric_limits_test.cpp:(.text+0x18a9): undefined reference to `std::numeric_limits<stan::math::fvar<stan::math::fvar<double> > >::has_signaling_NaN'
std_numeric_limits_test.cpp:(.text+0x1978): undefined reference to `std::numeric_limits<stan::math::fvar<double> >::has_denorm'
std_numeric_limits_test.cpp:(.text+0x1a47): undefined reference to `std::numeric_limits<stan::math::fvar<stan::math::fvar<double> > >::has_denorm'
std_numeric_limits_test.cpp:(.text+0x2034): undefined reference to `std::numeric_limits<stan::math::fvar<double> >::is_iec559'
std_numeric_limits_test.cpp:(.text+0x2103): undefined reference to `std::numeric_limits<stan::math::fvar<stan::math::fvar<double> > >::is_iec559'
std_numeric_limits_test.cpp:(.text+0x21d2): undefined reference to `std::numeric_limits<stan::math::fvar<double> >::is_bounded'
std_numeric_limits_test.cpp:(.text+0x22a1): undefined reference to `std::numeric_limits<stan::math::fvar<stan::math::fvar<double> > >::is_bounded'
std_numeric_limits_test.cpp:(.text+0x2862): undefined reference to `std::numeric_limits<stan::math::fvar<double> >::round_style'
std_numeric_limits_test.cpp:(.text+0x2931): undefined reference to `std::numeric_limits<stan::math::fvar<stan::math::fvar<double> > >::round_style'
collect2: error: ld returned 1 exit status
make/tests:16: recipe for target 'test/unit/math/fwd/core/std_numeric_limits_test' failed
make: *** [test/unit/math/fwd/core/std_numeric_limits_test] Error 1
make -j24 test/unit/math/fwd/core/std_numeric_limits_test failed
exit now (03/28/18 19:05:20 EDT)
```

#### Current Version:
v2.17.0

--- a/stan/math/fwd/core/std_numeric_limits.hpp
+++ b/stan/math/fwd/core/std_numeric_limits.hpp
@@ -8,15 +8,15 @@ namespace std {
 
 template <typename T>
 struct numeric_limits<stan::math::fvar<T> > {
-  static const bool is_specialized = true;
+  static const bool is_specialized;
   static stan::math::fvar<T> min() { return numeric_limits<double>::min(); }
   static stan::math::fvar<T> max() { return numeric_limits<double>::max(); }
-  static const int digits = numeric_limits<double>::digits;
-  static const int digits10 = numeric_limits<double>::digits10;
-  static const bool is_signed = numeric_limits<double>::is_signed;
-  static const bool is_integer = numeric_limits<double>::is_integer;
-  static const bool is_exact = numeric_limits<double>::is_exact;
-  static const int radix = numeric_limits<double>::radix;
+  static const int digits;
+  static const int digits10;
+  static const bool is_signed;
+  static const bool is_integer;
+  static const bool is_exact;
+  static const int radix;
   static stan::math::fvar<T> epsilon() {
     return numeric_limits<double>::epsilon();
   }
@@ -24,18 +24,17 @@ struct numeric_limits<stan::math::fvar<T> > {
     return numeric_limits<double>::round_error();
   }
 
-  static const int min_exponent = numeric_limits<double>::min_exponent;
-  static const int min_exponent10 = numeric_limits<double>::min_exponent10;
-  static const int max_exponent = numeric_limits<double>::max_exponent;
-  static const int max_exponent10 = numeric_limits<double>::max_exponent10;
+  static const int min_exponent;
+  static const int min_exponent10;
+  static const int max_exponent;
+  static const int max_exponent10;
 
-  static const bool has_infinity = numeric_limits<double>::has_infinity;
-  static const bool has_quiet_NaN = numeric_limits<double>::has_quiet_NaN;
-  static const bool has_signaling_NaN
-      = numeric_limits<double>::has_signaling_NaN;
-  static const float_denorm_style has_denorm
-      = numeric_limits<double>::has_denorm;
-  static const bool has_denorm_loss = numeric_limits<double>::has_denorm_loss;
+  static const bool has_infinity;
+  static const bool has_quiet_NaN;
+  static const bool has_signaling_NaN;
+
+  static const float_denorm_style has_denorm;
+  static const bool has_denorm_loss;
   static stan::math::fvar<T> infinity() {
     return numeric_limits<double>::infinity();
   }
@@ -49,14 +48,68 @@ struct numeric_limits<stan::math::fvar<T> > {
     return numeric_limits<double>::denorm_min();
   }
 
-  static const bool is_iec559 = numeric_limits<double>::is_iec559;
-  static const bool is_bounded = numeric_limits<double>::is_bounded;
-  static const bool is_modulo = numeric_limits<double>::is_modulo;
+  static const bool is_iec559;
+  static const bool is_bounded;
+  static const bool is_modulo;
 
-  static const bool traps = numeric_limits<double>::traps;
-  static const bool tinyness_before = numeric_limits<double>::tinyness_before;
-  static const float_round_style round_style
-      = numeric_limits<double>::round_style;
+  static const bool traps;
+  static const bool tinyness_before;
+  static const float_round_style round_style;
 };
+
+template <typename T>
+  const bool numeric_limits<stan::math::fvar<T> >::is_specialized = true;
+template <typename T>
+  const int numeric_limits<stan::math::fvar<T> >::digits = numeric_limits<double>::digits;
+template <typename T>
+  const int numeric_limits<stan::math::fvar<T> >::digits10 = numeric_limits<double>::digits10;
+template <typename T>
+  const bool numeric_limits<stan::math::fvar<T> >::is_signed = numeric_limits<double>::is_signed;
+template <typename T>
+  const bool numeric_limits<stan::math::fvar<T> >::is_integer = numeric_limits<double>::is_integer;
+template <typename T>
+  const bool numeric_limits<stan::math::fvar<T> >::is_exact = numeric_limits<double>::is_exact;
+template <typename T>
+  const int numeric_limits<stan::math::fvar<T> >::radix = numeric_limits<double>::radix;
+template <typename T>
+  const int numeric_limits<stan::math::fvar<T> >::min_exponent = numeric_limits<double>::min_exponent;
+template <typename T>
+  const int numeric_limits<stan::math::fvar<T> >::min_exponent10 = numeric_limits<double>::min_exponent10;
+template <typename T>
+  const int numeric_limits<stan::math::fvar<T> >::max_exponent = numeric_limits<double>::max_exponent;
+template <typename T>
+  const int numeric_limits<stan::math::fvar<T> >::max_exponent10 = numeric_limits<double>::max_exponent10;
+
+
+
+template <typename T>
+  const bool numeric_limits<stan::math::fvar<T> >::has_infinity = numeric_limits<double>::has_infinity;
+template <typename T>
+  const bool numeric_limits<stan::math::fvar<T> >::has_quiet_NaN = numeric_limits<double>::has_quiet_NaN;
+template <typename T>
+  const bool numeric_limits<stan::math::fvar<T> >::has_signaling_NaN = numeric_limits<double>::has_signaling_NaN;
+template <typename T>
+  const float_denorm_style numeric_limits<stan::math::fvar<T> >::has_denorm = numeric_limits<double>::has_denorm;
+template <typename T>
+  const bool numeric_limits<stan::math::fvar<T> >::has_denorm_loss = numeric_limits<double>::has_denorm_loss;
+
+template <typename T>
+  const bool numeric_limits<stan::math::fvar<T> >::is_iec559
+    = numeric_limits<double>::is_iec559;
+template <typename T>
+  const bool numeric_limits<stan::math::fvar<T> >::is_bounded
+    = numeric_limits<double>::is_bounded;
+template <typename T>
+  const bool numeric_limits<stan::math::fvar<T> >::is_modulo
+    = numeric_limits<double>::is_modulo;
+
+template <typename T>
+  const bool numeric_limits<stan::math::fvar<T> >::traps = numeric_limits<double>::traps;
+template <typename T>
+  const bool numeric_limits<stan::math::fvar<T> >::tinyness_before = numeric_limits<double>::tinyness_before;
+template <typename T>
+  const float_round_style numeric_limits<stan::math::fvar<T> >::round_style
+    = numeric_limits<double>::round_style;
+
 }  // namespace std
 #endif

--- a/stan/math/fwd/core/std_numeric_limits.hpp
+++ b/stan/math/fwd/core/std_numeric_limits.hpp
@@ -60,38 +60,53 @@ struct numeric_limits<stan::math::fvar<T> > {
 template <typename T>
   const bool numeric_limits<stan::math::fvar<T> >::is_specialized = true;
 template <typename T>
-  const int numeric_limits<stan::math::fvar<T> >::digits = numeric_limits<double>::digits;
+  const int numeric_limits<stan::math::fvar<T> >::digits
+    = numeric_limits<double>::digits;
 template <typename T>
-  const int numeric_limits<stan::math::fvar<T> >::digits10 = numeric_limits<double>::digits10;
+  const int numeric_limits<stan::math::fvar<T> >::digits10
+    = numeric_limits<double>::digits10;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::is_signed = numeric_limits<double>::is_signed;
+  const bool numeric_limits<stan::math::fvar<T> >::is_signed
+    = numeric_limits<double>::is_signed;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::is_integer = numeric_limits<double>::is_integer;
+  const bool numeric_limits<stan::math::fvar<T> >::is_integer
+    = numeric_limits<double>::is_integer;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::is_exact = numeric_limits<double>::is_exact;
+  const bool numeric_limits<stan::math::fvar<T> >::is_exact
+    = numeric_limits<double>::is_exact;
 template <typename T>
-  const int numeric_limits<stan::math::fvar<T> >::radix = numeric_limits<double>::radix;
+  const int numeric_limits<stan::math::fvar<T> >::radix
+    = numeric_limits<double>::radix;
 template <typename T>
-  const int numeric_limits<stan::math::fvar<T> >::min_exponent = numeric_limits<double>::min_exponent;
+  const int numeric_limits<stan::math::fvar<T> >::min_exponent
+    = numeric_limits<double>::min_exponent;
 template <typename T>
-  const int numeric_limits<stan::math::fvar<T> >::min_exponent10 = numeric_limits<double>::min_exponent10;
+  const int numeric_limits<stan::math::fvar<T> >::min_exponent10
+    = numeric_limits<double>::min_exponent10;
 template <typename T>
-  const int numeric_limits<stan::math::fvar<T> >::max_exponent = numeric_limits<double>::max_exponent;
+  const int numeric_limits<stan::math::fvar<T> >::max_exponent
+    = numeric_limits<double>::max_exponent;
 template <typename T>
-  const int numeric_limits<stan::math::fvar<T> >::max_exponent10 = numeric_limits<double>::max_exponent10;
+  const int numeric_limits<stan::math::fvar<T> >::max_exponent10
+    = numeric_limits<double>::max_exponent10;
 
 
 
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::has_infinity = numeric_limits<double>::has_infinity;
+  const bool numeric_limits<stan::math::fvar<T> >::has_infinity
+    = numeric_limits<double>::has_infinity;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::has_quiet_NaN = numeric_limits<double>::has_quiet_NaN;
+  const bool numeric_limits<stan::math::fvar<T> >::has_quiet_NaN
+    = numeric_limits<double>::has_quiet_NaN;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::has_signaling_NaN = numeric_limits<double>::has_signaling_NaN;
+  const bool numeric_limits<stan::math::fvar<T> >::has_signaling_NaN
+    = numeric_limits<double>::has_signaling_NaN;
 template <typename T>
-  const float_denorm_style numeric_limits<stan::math::fvar<T> >::has_denorm = numeric_limits<double>::has_denorm;
+  const float_denorm_style numeric_limits<stan::math::fvar<T> >::has_denorm
+    = numeric_limits<double>::has_denorm;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::has_denorm_loss = numeric_limits<double>::has_denorm_loss;
+  const bool numeric_limits<stan::math::fvar<T> >::has_denorm_loss
+    = numeric_limits<double>::has_denorm_loss;
 
 template <typename T>
   const bool numeric_limits<stan::math::fvar<T> >::is_iec559
@@ -104,9 +119,11 @@ template <typename T>
     = numeric_limits<double>::is_modulo;
 
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::traps = numeric_limits<double>::traps;
+  const bool numeric_limits<stan::math::fvar<T> >::traps
+    = numeric_limits<double>::traps;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::tinyness_before = numeric_limits<double>::tinyness_before;
+  const bool numeric_limits<stan::math::fvar<T> >::tinyness_before
+    = numeric_limits<double>::tinyness_before;
 template <typename T>
   const float_round_style numeric_limits<stan::math::fvar<T> >::round_style
     = numeric_limits<double>::round_style;

--- a/stan/math/fwd/core/std_numeric_limits.hpp
+++ b/stan/math/fwd/core/std_numeric_limits.hpp
@@ -58,74 +58,72 @@ struct numeric_limits<stan::math::fvar<T> > {
 };
 
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::is_specialized = true;
+const bool numeric_limits<stan::math::fvar<T> >::is_specialized = true;
 template <typename T>
-  const int numeric_limits<stan::math::fvar<T> >::digits
+const int numeric_limits<stan::math::fvar<T> >::digits
     = numeric_limits<double>::digits;
 template <typename T>
-  const int numeric_limits<stan::math::fvar<T> >::digits10
+const int numeric_limits<stan::math::fvar<T> >::digits10
     = numeric_limits<double>::digits10;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::is_signed
+const bool numeric_limits<stan::math::fvar<T> >::is_signed
     = numeric_limits<double>::is_signed;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::is_integer
+const bool numeric_limits<stan::math::fvar<T> >::is_integer
     = numeric_limits<double>::is_integer;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::is_exact
+const bool numeric_limits<stan::math::fvar<T> >::is_exact
     = numeric_limits<double>::is_exact;
 template <typename T>
-  const int numeric_limits<stan::math::fvar<T> >::radix
+const int numeric_limits<stan::math::fvar<T> >::radix
     = numeric_limits<double>::radix;
 template <typename T>
-  const int numeric_limits<stan::math::fvar<T> >::min_exponent
+const int numeric_limits<stan::math::fvar<T> >::min_exponent
     = numeric_limits<double>::min_exponent;
 template <typename T>
-  const int numeric_limits<stan::math::fvar<T> >::min_exponent10
+const int numeric_limits<stan::math::fvar<T> >::min_exponent10
     = numeric_limits<double>::min_exponent10;
 template <typename T>
-  const int numeric_limits<stan::math::fvar<T> >::max_exponent
+const int numeric_limits<stan::math::fvar<T> >::max_exponent
     = numeric_limits<double>::max_exponent;
 template <typename T>
-  const int numeric_limits<stan::math::fvar<T> >::max_exponent10
+const int numeric_limits<stan::math::fvar<T> >::max_exponent10
     = numeric_limits<double>::max_exponent10;
 
-
-
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::has_infinity
+const bool numeric_limits<stan::math::fvar<T> >::has_infinity
     = numeric_limits<double>::has_infinity;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::has_quiet_NaN
+const bool numeric_limits<stan::math::fvar<T> >::has_quiet_NaN
     = numeric_limits<double>::has_quiet_NaN;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::has_signaling_NaN
+const bool numeric_limits<stan::math::fvar<T> >::has_signaling_NaN
     = numeric_limits<double>::has_signaling_NaN;
 template <typename T>
-  const float_denorm_style numeric_limits<stan::math::fvar<T> >::has_denorm
+const float_denorm_style numeric_limits<stan::math::fvar<T> >::has_denorm
     = numeric_limits<double>::has_denorm;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::has_denorm_loss
+const bool numeric_limits<stan::math::fvar<T> >::has_denorm_loss
     = numeric_limits<double>::has_denorm_loss;
 
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::is_iec559
+const bool numeric_limits<stan::math::fvar<T> >::is_iec559
     = numeric_limits<double>::is_iec559;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::is_bounded
+const bool numeric_limits<stan::math::fvar<T> >::is_bounded
     = numeric_limits<double>::is_bounded;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::is_modulo
+const bool numeric_limits<stan::math::fvar<T> >::is_modulo
     = numeric_limits<double>::is_modulo;
 
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::traps
+const bool numeric_limits<stan::math::fvar<T> >::traps
     = numeric_limits<double>::traps;
 template <typename T>
-  const bool numeric_limits<stan::math::fvar<T> >::tinyness_before
+const bool numeric_limits<stan::math::fvar<T> >::tinyness_before
     = numeric_limits<double>::tinyness_before;
 template <typename T>
-  const float_round_style numeric_limits<stan::math::fvar<T> >::round_style
+const float_round_style numeric_limits<stan::math::fvar<T> >::round_style
     = numeric_limits<double>::round_style;
 
 }  // namespace std

